### ISSUE TITLE
Add way to build project with `-fno-emit-bin` flag set for error reporting

### ISSUE THREAD

**Issue by Jarred-Sumner** [zigtools/zls#870]
This would fix https://github.com/zigtools/zls/issues/869 and https://github.com/zigtools/zls/issues/868 in a maintainable way and bring the Zig developer experience closer to what mature programming languages like Rust or Go have.

**Comment by SuperAuguste** [comment]
@sin-ack was talking about this in the Zig Discord and I think it's a great idea, there are just a few performance things we need to consider. This is probably something that should be implemented with threads + request cancellation/continuation.

**Comment by Jarred-Sumner** [comment]
`-fno-emit-bin` should be fast enough for most projects. For bun, it's 8 seconds on an M1 (16 seconds on linux). That is too long, but Bun is one of the largest Zig projects so it should be a lot faster for everyone else.

If you use an event loop + posix_spawn and then disable the filesystem locking for zig cache (which prevents multiple instances to run), it should work well enough. posix_spawn is faster when the process uses lots of memory. The event loop should monitor file descriptor readiness so you can call `waitpid` without blocking the thread. This design would let you keep zls as one thread and start multiple instances of `zig build` without waiting for previous ones to complete.  If there is no easy way to prevent file locking, maybe just have a pool of up to 8 or so instances of `zig build` each with 8 different caches and cycle through each one

**Comment by SuperAuguste** [comment]
With that benchmark in mind this actually makes sense to do, I'll try my hand at hacking away at it :)

**Comment by Jarred-Sumner** [comment]
If you want some sample code which does the posix spawn waitpid stuff (note that this is platform-specific code for Linux and macOS)

- https://github.com/oven-sh/bun/blob/b86915c80819e61e6e1932d7802a4c31c73094e9/src/bun.js/api/bun/subprocess.zig#L1472-L1491
- https://github.com/oven-sh/bun/blob/b86915c80819e61e6e1932d7802a4c31c73094e9/src/bun.js/api/bun/subprocess.zig#L1391-L1408
- https://github.com/oven-sh/bun/blob/b86915c80819e61e6e1932d7802a4c31c73094e9/src/bun.js/base.zig#L3412-L3454
- https://github.com/oven-sh/bun/blob/b86915c80819e61e6e1932d7802a4c31c73094e9/src/bun.js/api/bun/subprocess.zig#L1273-L1282

**Comment by SuperAuguste** [comment]
I think I'll just get something simple and platform independent working first and then I might look at optimizing it if it's really outrageously slow or something :P

**Comment by SuperAuguste** [comment]
![image](https://user-images.githubusercontent.com/19855629/210130424-551b9025-4014-4dd5-9803-e4c55cf70dde.png)

Tada! Very rudimentary and it's pretty slow but at least it works :)

**Comment by SuperAuguste** [comment]
So I implemented this in its hacky asynchronous entirety (on Windows at least) and sadly some concerns I had turned out to be true: it is straight up unusable. I'm testing this on a fresh `zig init-exe` and it takes a good 5 seconds between saves which is already pretty bad (this is with all emits off btw). The experience is then worsened by the fact that whenever I try and fix an error by modifying my code, all the indices are all completely off.

https://user-images.githubusercontent.com/19855629/210184310-ba5ab439-bc72-43af-9574-c156fd00a550.mp4

Sadly, I think this makes this feature wholly useless for most people :(
